libraries/integration/Microsoft.Bot.Builder.Integration.Runtime/Extensions/ServiceCollectionExtensions.cs

```diff
@@ -0,0 +1,235 @@
+
+            // Configuration
+            string applicationRoot = configuration.GetSection(ConfigurationConstants.ApplicationRootKey).Value;
+            string defaultLocale = configuration.GetSection(ConfigurationConstants.DefaultLocale).Value;
+            string rootDialog = configuration.GetSection(ConfigurationConstants.RootDialogKey).Value;
+            
+            // Runtime settings. If no config is provided, we create basic runtime config with defaults.
+            var runtimeSettings = configuration.GetSection(ConfigurationConstants.RuntimeSettingsKey).Get<RuntimeSettings>() ?? new RuntimeSettings();
+
+            // Bot
+            services.AddSingleton<IBot, CoreBot>();
+            services.AddOptions()
+                .Configure<CoreBotOptions>(o =>
+                {
+                    o.DefaultLocale = defaultLocale;
+                    o.RootDialog = rootDialog;
+                });
+
+            services.AddSingleton(configuration);
+
+            // ResourceExplorer. TryAddSingleton will only add if there is no other registration for resource explorer.
+            // Tests use this to inject custom resource explorers but could also be used for advanced runtime customization scenarios.
+            services.TryAddSingleton<ResourceExplorer>(serviceProvider =>
+                new ResourceExplorer()
+                    .AddFolder(applicationRoot)
+                    .RegisterType<OnQnAMatch>(OnQnAMatch.Kind));
+            
+            // Runtime set up
+            services.AddBotRuntimeSkills(runtimeSettings.Skills);
+            services.AddBotRuntimeStorage(configuration, runtimeSettings);
+            services.AddBotRuntimeTelemetry(runtimeSettings.Telemetry);
+            services.AddBotRuntimeTranscriptLogging(configuration, runtimeSettings.Features);
+            services.AddBotRuntimeFeatures(runtimeSettings.Features);
+            services.AddBotRuntimePlugins(configuration, runtimeSettings);
+            services.AddBotRuntimeAdapters(runtimeSettings);
+        }
+
+        internal static void AddBotRuntimeSkills(this IServiceCollection services, SkillSettings skillSettings)
+        {
+            services.AddSingleton(sp => new AuthenticationConfiguration { ClaimsValidator = new AllowedCallersClaimsValidator(skillSettings?.AllowedCallers) });
+            services.AddSingleton<SkillConversationIdFactoryBase, SkillConversationIdFactory>();
+            services.AddHttpClient<BotFrameworkClient, SkillHttpClient>();
+            services.AddSingleton<ChannelServiceHandler, SkillHandler>();
+        }
+
+        internal static void AddBotRuntimeStorage(this IServiceCollection services, IConfiguration configuration, RuntimeSettings runtimeSettings)
+        {
+            services.AddSingleton<UserState>();
+            services.AddSingleton<ConversationState>();
+
+            // Cosmosdb
+            if (runtimeSettings?.Storage == typeof(CosmosDbPartitionedStorage).Name)
```

Comment:
>typeof(CosmosDbPartitionedStorage).Name [](start = 44, length = 39)

Can use nameof()

Reply:
In this case we want to get the name of the type and unfortuanately nameof cannot be used on types, just on variables. Would have been nice, though.

Reply:
@carlosscastro  How about IsAssignableFrom?

Also, is there any way to use different storage providers for UserState vs ConversationState, or logging?

Reply:
I got this wrong, fixed. @EricDahlvang you can `AddService<UserState>(sp => new UserState(new PotatoStorage))` in your startup.cs.

There is something subtle about this though, which is for these things that we'll document overriding, we should probably use TryAddSingleton<T> in our own code instead of AddSingleton which removes the ordering dependency and is a way of saying 'This is the default, in case nobody registers'. I'm testing some more ways of doing this and the final nice way might land in R13 but there should be clearly documented hooks that if you register yourself will take precedence. I'll add some tests for your multi-storage example next week.